### PR TITLE
feat!: Add informative response fields for leverage messages

### PR DESCRIPTION
**Pull request by toteki**
Main changes:
- Leverage `Message` types now return any `sdk.Coins` representing effects achieved that are not equal to input fields. This includes uTokens received when supplying, and base tokens received when withdrawing.

And some miscellaneous changes:
- Reorder `make proto-all` for ideal execution. (Format first, then lint, then generate, then swagger)
- Standardized `Coin` -> `Asset` in `MsgCollateralize` and `MsgDecollateralize`
- Event `repay_borrowed_asset` is now just `repay`
- Event `supply` attribute `amount` removed, now has `supplied` and `received`
- Event `withdraw` attribute `amount` removed, now has `redeemed` and `received`

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] added appropriate labels to the PR
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [x] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/leverage/keeper/keeper_test.go [hunk 1]
```diff
@@ -126,7 +126,7 @@ func (s *IntegrationTestSuite) setupAccount(denom string, mintAmount, supplyAmou
```
Comment (toteki) [inline x/leverage/keeper/keeper_test.go:126]: These new return fields can be checked to improve our test suite in the future, but I want to hold off on that until I refactor `x/leverage/keeper` tests.

For now, test behavior is unchanged.

File: proto/umee/leverage/v1/tx.proto [hunk 2]
```diff
@@ -82,57 +82,75 @@ message MsgRepay {
...
   // will be taken from the borrower's collateral.
-  string                   reward_denom = 4;
+  string reward_denom = 4;
 }
 
 // MsgSupplyResponse defines the Msg/Supply response type.
-message MsgSupplyResponse {}
+message MsgSupplyResponse {
+  // Supplied is the amount of base tokens sent ot the module.
+  cosmos.base.v1beta1.Coin supplied = 1 [(gogoproto.nullable) = false];
```
Comment (robert-zaremba) [inline proto/umee/leverage/v1/tx.proto:108]: this attribute is not needed - it's a copy of the input (msg.Asset). No need to record it (it's going to be saved in tendermint store)

File: proto/umee/leverage/v1/tx.proto [hunk 3]
```diff
@@ -82,57 +82,75 @@ message MsgRepay {
...
+  // Received is the amount of uTokens received.
+  cosmos.base.v1beta1.Coin received = 2 [(gogoproto.nullable) = false];
+}
 
 // MsgWithdrawResponse defines the Msg/Withdraw response type.
-message MsgWithdrawResponse {}
+message MsgWithdrawResponse {
+  // Redeemed is the amount of uTokens sent to the module.
+  cosmos.base.v1beta1.Coin redeemed = 1 [(gogoproto.nullable) = false];
```
Comment (robert-zaremba) [inline proto/umee/leverage/v1/tx.proto:116]: ditto

File: proto/umee/leverage/v1/tx.proto [hunk 4]
```diff
@@ -82,57 +82,75 @@ message MsgRepay {
...
+  // Received is the amount of base tokens received.
+  cosmos.base.v1beta1.Coin received = 2 [(gogoproto.nullable) = false];
+}
 
 // MsgCollateralizeResponse defines the Msg/Collateralize response type.
-message MsgCollateralizeResponse {}
+message MsgCollateralizeResponse {
+  // Collateralized is the amount of uTokens sent to the module.
+  cosmos.base.v1beta1.Coin collateralized = 1 [(gogoproto.nullable) = false];
```
Comment (robert-zaremba) [inline proto/umee/leverage/v1/tx.proto:124]: ditto, no need to record this.

File: proto/umee/leverage/v1/tx.proto [hunk 5]
```diff
@@ -82,57 +82,75 @@ message MsgRepay {
...
+message MsgDecollateralizeResponse {
+  // Received is the amount of uTokens retrieved from the module.
+  cosmos.base.v1beta1.Coin received = 1 [(gogoproto.nullable) = false];
+}
 
 // MsgBorrowResponse defines the Msg/Borrow response type.
-message MsgBorrowResponse {}
+message MsgBorrowResponse {
+  // Borrowed is the amount of uTokens borrowed from the module.
+  cosmos.base.v1beta1.Coin borrowed = 1 [(gogoproto.nullable) = false];
```
Comment (robert-zaremba) [inline proto/umee/leverage/v1/tx.proto:136]: same here

File: x/leverage/keeper/keeper_test.go [hunk 6]
```diff
@@ -126,7 +126,7 @@ func (s *IntegrationTestSuite) setupAccount(denom string, mintAmount, supplyAmou
 
 	if supplyAmount > 0 {
 		// account supplies supplyAmount tokens and receives uTokens
-		err := s.app.LeverageKeeper.Supply(s.ctx, addr, sdk.NewInt64Coin(denom, supplyAmount))
+		_, err := s.app.LeverageKeeper.Supply(s.ctx, addr, sdk.NewInt64Coin(denom, supplyAmount))
```
Comment (robert-zaremba) [inline x/leverage/keeper/keeper_test.go:129]: let's test the response `Received` value.

**Review by robert-zaremba — COMMENTED** [review]
many good updates.

- [x] let's don't store in response a copy of input - no need to pollute storage
- [ ] proto messages should follow the API language - we decided to use Assest rather than coin / token. If we want to change the language of API, then it's better to firstly discuss it on Slack, rather than in PR (unless something is small) to avoid going too much back and forth.

File: proto/umee/leverage/v1/tx.proto [hunk 7]
```diff
@@ -40,15 +40,15 @@ service Msg {
 message MsgSupply {
   // Supplier is the account address supplying assets and the signer of the message.
   string                   supplier = 1;
-  cosmos.base.v1beta1.Coin asset    = 2 [(gogoproto.nullable) = false];
+  cosmos.base.v1beta1.Coin coin     = 2 [(gogoproto.nullable) = false];
```
[the remark this answers is not in the document]
Comment (toteki) [reply proto/umee/leverage/v1/tx.proto:43]: I thought we did the opposite - might have misremembered, but I can revert that

[the remark this answers is not in the document]
Comment (robert-zaremba) [reply proto/umee/leverage/v1/tx.proto:43]: Hmm, i'm getting confused as well ;) 
https://github.com/umee-network/umee/issues/1102 -- here we use coin
But in documentation we use both, asset and coin, see:
+ design_docs/002-deposit-assets.md vs 
+ design_docs/009-liquidity-mining.md
Asset is used more often. We there was also `MsgLendAsset` (now `MsgSupply`)

I think the idea was to use `Asset` in proto, but `coin` can be used in local functions as an argument name. 

What do you think?

File: CHANGELOG.md [hunk 8]
```diff
@@ -79,6 +81,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 - [1203](https://github.com/umee-network/umee/pull/1203) Add Swagger docs.
 - [1212](https://github.com/umee-network/umee/pull/1212) Add `util/checkers` utility package providing common check / validation functions.
 - [1220](https://github.com/umee-network/umee/pull/1220) Submit oracle prevotes / vote txs via the CLI.
+- [1236](https://github.com/umee-network/umee/pull/1236) Improve leverage event fields.
```
[the remark this answers is not in the document]
Comment (toteki) [reply CHANGELOG.md]: done

File: proto/umee/leverage/v1/tx.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (toteki) [reply proto/umee/leverage/v1/tx.proto:43]: That sounds good. I'll go look for for occurrences of "asset" in the codebase to be sure.

File: Makefile [hunk 9]
```diff
@@ -200,8 +200,8 @@ containerProtoGen=$(PROJECT_NAME)-proto-gen-$(containerProtoVer)
 containerProtoFmt=$(PROJECT_NAME)-proto-fmt-$(containerProtoVer)
 containerProtoGenSwagger=$(PROJECT_NAME)-proto-gen-swagger-$(containerProtoVer)
 
-proto-all: proto-gen proto-lint proto-check-breaking proto-format
-.PHONY: proto-all proto-gen proto-lint proto-check-breaking proto-format
+proto-all: proto-format proto-lint proto-gen proto-swagger-gen proto-check-breaking
```
[the remark this answers is not in the document]
Comment (toteki) [reply Makefile]: that works

File: proto/umee/leverage/v1/tx.proto (hunk 2 above)
Comment (toteki) [reply proto/umee/leverage/v1/tx.proto]: Noted. Removed all input-matching response fields

**Comment by toteki** [comment]
Note: I intend to write some newly possible tests in `keeper_test.go`, so no need to look at this again until then. Other files are complete though.

File: x/leverage/keeper/msg_server.go [hunk 10]
```diff
@@ -31,21 +31,24 @@ func (s msgServer) Supply(
...
+		"received", received.String(),
 	)
 
 	ctx.EventManager().EmitEvents(sdk.Events{
 		sdk.NewEvent(
 			types.EventTypeSupply,
 			sdk.NewAttribute(types.EventAttrSupplier, supplierAddr.String()),
-			sdk.NewAttribute(sdk.AttributeKeyAmount, msg.Asset.String()),
+			sdk.NewAttribute(types.EventAttrSupplied, msg.Asset.String()),
+			sdk.NewAttribute(types.EventAttrReceived, received.String()),
```
Comment (robert-zaremba) [inline x/leverage/keeper/msg_server.go:51]: side note: with new features in 0.46 (responses registered on chain), the events have less sense.

**Review by robert-zaremba — APPROVED** [review]
Pre-approving. 
- [ ] please add at least one test with a new response
- [ ] create an issue to add more tests.

File: CHANGELOG.md [hunk 11]
```diff
@@ -86,21 +87,19 @@ Ref: https://keepachangelog.com/en/1.0.0/
...
 
 ### Bug Fixes
 
 - [1018](https://github.com/umee-network/umee/pull/1018) Return nil if negative time elapsed from the last block happens.
 - [1156](https://github.com/umee-network/umee/pull/1156) Propagate context correctly.
 
-### API Breaking
-
-- [926](https://github.com/umee-network/umee/pull/926)(x/leverage) Renamed `Keeper.DeriveBorrowUtilization` to `SupplyUtilization`.
-
```
Comment (toteki) [inline CHANGELOG.md:103]: Note: moved to improvements section and linked PR (940) instead of issue (926).

**Comment by toteki** [comment]
Initial tests added. Test improvement issue created at #1242

**Pull request merged**